apps/trackers/jira/query.py

```diff
@@ -267,3 +271,89 @@ def generate_additional_fields(self):
                     continue
 
                 self._query["fields"][PS_ADDITIONAL_FIELD_TO_JIRA[name]] = field_value
+
+    def generate_cc(self):
+        """
+        generate query for CC list
+        """
+
+        # Each instance of TrackerJiraQueryBuilder is used only once, but if ever used twice,
+        # always produce consistent query and comment.
+        self._comment = None
+
+        if self.tracker.external_system_id:
+            # Add CCs only on creation.
+            return
+
+        # NOTE That SFM2 for Jira tracker creation uses ps_module.component_overrides only for
+        #      generating Jira "components" field, which OSIDB doesn't do, but not for CC lists;
+        #      CC list creation is based solely on ps_component, not on bz_component.
+        #      Therefore AffectCCBuilder.ps2bz_component is not reused here.
+        #      TODO: Is this a bug?
+        # Parse BZ component
+        if self.ps_component and "/" in self.ps_component:
+            bz_component = self.ps_component.split("/")[-1]
+        else:
+            bz_component = self.ps_component
+
+        # embargoed value unused here
+        affect_cc_builder = AffectCCBuilder(
+            self.tracker.affects.first(), embargoed=None
+        )
+        # TODO: Why does AffectCCBuilder set bz_component to None?
+        affect_cc_builder.bz_component = bz_component
+        cc_list = set(affect_cc_builder.component_cc())
+
+        if self.ps_module.default_cc:
+            # Default CC List
+            cc_list.update(self.ps_module.default_cc)
+        if self.tracker.is_embargoed and self.ps_module.private_tracker_cc:
+            cc_list.update(self.ps_module.private_tracker_cc)
+
+        # Replaces contact aliases with appropriate emails/usernames if alias
+        # has the contact set for current BTS. Other records are made intact.
+        # NOTE: Similar functionality in AffectCCBuilder.expand_alias(),
+        #       but only for a single contact at a time.
+        contacts = dict(
+            PsContact.objects.all()
+            .values_list("username", "jboss_username")
+            .filter(username__in=cc_list)
+        )
+        cc_list = {contacts.get(cc, cc) for cc in cc_list}
```

Comment:
I would be for extending the `expand_alias` method to support the bulk expansion rather than implementing the expansion on two different places.

Reply:
It is too complicated indeed. But I think that the significant part of the complicated matter is that SFM2 code is a complete mess. There are patterns and building blocks I tried to identify, eg.

- flaw CC is composition affect CCs
- affect CC is composition of module, component, and Bugzilla CCs
- RHSCL affect CC is affect CC with a different component handling
- etc.

so I think that decomposition and deduplication is valuable here. I think this code will live with us for some time and I wanted to avoid repeating the SFM2 unmaintainable mess - we might be soon able to cut the sync from Bugzilla but the other direction does not seem easy to be thrown away.

However, I do not want to dictate it here. If you think that deduplication would be too complicated or too risky in the sense of introducing new errors then do it somehow so it done :smiley: